keps/sig-network/2009-dns-autopath/README.md

```diff
@@ -0,0 +1,228 @@
+
+1) The query will be sent out as `mytestsvc.search.default.cluster.local.ap.k8s.io`. 
+
+2) clusterDNS server(coreDNS by default) receives this and strips off the delimiter - "ap.k8s.io" and identifies "search" as start of the custom searchpath. The namespace and cluster suffix can be obtained from rest of the string. The DNS server can now construct the full service name for this query as - `mytestsvc.default.svc.cluster.local`. 
+
+This approach minimizes the number of DNS queries at client side to atmost 2(A, AAAA). The searchpath expansion logic moves to the server side.
+
+However, the above approach requires the clusterDNS server to know about the k8s DNS schema as well as the syntax of this custom searchpath. It does know about k8s DNS schema anyway, but knowledge of the custom searchpath is an additional requirement.
+
+### New EDNS0 Option
+
+As an alternative, we introduce an EDNS0 option which will include a list of
+searchpaths to be applied to the base query. 
+
+A new EDNS0 option - SearchPaths is introduced, with an option code from the [experimental range - 65001 to 65534](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-11).
+The value of this option will be a comma-separated string consisting of all the searchpaths which are to be appended to the main query and looked up. This option can be useful outside of Kubernetes as well.
+
+
+Instead of having to modify all client pod images to insert an EDNS0 option in their requests, a new CoreDNS plugin "gensearchpaths" will be introduced. This will generate a list of search paths as an EDNS0 option to a given query.
+
+This plugin can be used by [Nodelocal DNSCache](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0030-nodelocal-dns-cache.md), which is a daemonset running a per-node DNS Cache. It can also be used by a sidecar container attached to a pod, to selectively use this feature. The sidecar container will run a stripped-down version of CoreDNS with the above plugin.
+
+
+The clusterDNS service needs to support this new EDNS0 option and lookup multiple query names by expanding a single incoming query. This was tried on a test setup by [modifying the autopath plugin](https://github.com/coredns/coredns/compare/master...prameshj:auto) in CoreDNS to extract the searchpath from an EDNS0 option, for a proof of concept.
+
+If upstream ClusterDNS uses something other than CoreDNS, support for this EDNS0 option should be added for autopath to work. 
+
+### Risks and Mitigations
+
+1) If NodeLocal DNSCache is responsible for adding the EDNS0 option, DNS resolution can break if NodeLocal DNSCache is down or if the pods configured with the special searchpath point to the kube-dns service directly to resolve query names. This is because without the EDNS0 option, the custom searchpath is not resolvable by kube-dns/CoreDNS. Running 2 DNSCache instances would be necessary to keep searchpath expansion working during upgrades.
```

Comment:
This is a significant limitation, IMO, and could make it infeasible to rely on the proxy.